src/t_zset.c

```diff
@@ -1851,32 +1853,35 @@ static void zaddGenericCommand(client *c, int flags) {
         ele = c->argv[scoreidx + 1 + j * 2]->ptr;
         int retval = zsetAdd(zobj, score, ele, flags, &retflags, &newscore);
         if (retval == 0) {
-            addReplyError(c, nanerr);
-            goto cleanup;
+            reply_err = 1;
+            break;
         }
         if (retflags & ZADD_OUT_ADDED) added++;
         if (retflags & ZADD_OUT_UPDATED) updated++;
         if (!(retflags & ZADD_OUT_NOP)) processed++;
         score = newscore;
     }
-    server.dirty += (added + updated);
+    if (!reply_err) {
+        server.dirty += (added + updated);
+    }
+    if (added || updated) {
+        signalModifiedKey(c, c->db, key);
+        notifyKeyspaceEvent(c, NOTIFY_ZSET, incr ? "zincr" : "zadd", key, c->db->id);
```

Comment:
So now we need to notify event before addReply, right? We may need to add an assert in notifyKeyspaceEvent to check that the client has not replied.

And this event block cannot return any errors? we need to document it.

Reply:
I also like the idea of an assertion. I would only add it should probably be a debug assertion, since I don't really want to crash the server in production because of it.